DM-37868: fix more fallout from DM-31725

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

TallJimbo committed:
Remove apparently-unused test case variable.

I see no other references to this in daf_butler, pipe_base, obs_base,
ctrl_mpexec, or ctrl_bps, and at least my local MyPy wants it to be
annotated if it needs to stay; let's see if we can just get rid of it.

TallJimbo committed:
Type annotation fixes for MyPy 1.0.0.

TallJimbo committed:
Fix untyped defs and modernize type imports in config.py.

TallJimbo committed:
Remove problematic collection-type logic in findDatasets.

DM-31725 made it an error to search a mix of CALIBRATION- and
non-CALIBRATION collections for a dataset with no timespan, unless the
dataset was found in a non-CALIBRATION collection that appeared earlier
in the sequence than the first CALIBRATION collection, on the
assumption that it would guard against users forgetting to provide a
timespan and then being surprised that the CALIBRATION collection
wasn't searched at all (silently!).

But this breaks the standard workaround for searching for "unbounded"
calibration datasets, which is to include a special unbounded
non-CALIBRATION collection in the same chain, typically after one or
more CALIBRATION collections.

This commit removes this guard, causing findDatasets to silently ignore
all CALIBRATION collections when no timespan is provided.

TallJimbo committed:
Use property instead of already-closed connection for dialect.

TallJimbo committed:
Shrink identifiers for relation-based queries.

This keeps SQL identifiers coming out of daf_relation-based query
generation below PostgreSQL's 64-char limit.

TallJimbo committed:
Cache already-shrunk SQL names.

TallJimbo committed:
Add tests for too-long names in queries.

This is expected to cause failures in PostgreSQL.

TallJimbo committed:
Add changelog entry.

andy-slac commented on python/lsst/daf/butler/registry/tests/_registry.py:
```diff
@@ -3165,3 +3165,33 @@ def testExposureQueries(self):
+        longer than 64 chars, but it's worth testing for all DBs.
+        """
+        registry = self.makeRegistry()
+        name = "this_is_a_very_long_name_that_exceeds_the_postgresql_sixty_four_char_limit"
```
I'm going to believe you, but it's not easy to verify that this is actually longer than 64 chars.

TallJimbo replied:
I'll replace it.

andy-slac reviewed (approved):
Looks good. I can't parse what `login` means in `Remove problematic collection-type login in findDatasets` commit message.

TallJimbo commented:
> I can't parse what `login` means

Should have been "logic"; will fix.

TallJimbo committed:
Drop types-click dependency.

Click 8.x provides its own types, and the ones in types-click are for
Click 7.x and are hence incorrect.
